cns/restserver/ipam.go

```diff
@@ -478,101 +516,127 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+func (service *HTTPRestService) AssignAvailableIPConfigs(podInfo cns.PodInfo) ([]cns.PodIpInfo, error) {
 	service.Lock()
 	defer service.Unlock()
+	// Creates a slice of PoPodIpInfo with the size of number of NCs
+	podIPInfo := make([]cns.PodIpInfo, len(service.state.ContainerStatus))
+	// This map is used to store whether or not we have found an available IP from an NC when looping through the pool
+	ncMap := make(map[string]any)
 
 	for _, ipState := range service.PodIPConfigState {
-		if ipState.GetState() == types.Available {
+		_, found := ncMap[ipState.NCID]
+		// Checks if the current IP is available and if we haven't already found an IP from that NC
+		if !found && ipState.GetState() == types.Available {
 			if err := service.assignIPConfig(ipState, podInfo); err != nil {
-				return cns.PodIpInfo{}, err
+				logger.Errorf(err.Error())
+				break
 			}
 
-			podIPInfo := cns.PodIpInfo{}
-			if err := service.populateIPConfigInfoUntransacted(ipState, &podIPInfo); err != nil {
-				return cns.PodIpInfo{}, err
+			if err := service.populateIPConfigInfoUntransacted(ipState, &podIPInfo[len(ncMap)]); err != nil {
+				logger.Errorf(err.Error())
+				break
+			}
+			ncMap[ipState.NCID] = ipState
+			if len(ncMap) == len(service.state.ContainerStatus) {
+				return podIPInfo, nil
 			}
-
-			return podIPInfo, nil
 		}
 	}
+
 	//nolint:goerr113
-	return cns.PodIpInfo{}, fmt.Errorf("no IPs available, waiting on Azure CNS to allocate more")
+	return podIPInfo, fmt.Errorf("not enough IPs available, waiting on Azure CNS to allocate more")
 }
 
 // If IPConfig is already assigned to pod, it returns that else it returns one of the available ipconfigs.
-func requestIPConfigHelper(service *HTTPRestService, req cns.IPConfigRequest) (cns.PodIpInfo, error) {
+func requestIPConfigHelper(service *HTTPRestService, req cns.IPConfigRequest) ([]cns.PodIpInfo, error) {
 	// check if ipconfig already assigned tothis pod and return if exists or error
 	// if error, ipstate is nil, if exists, ipstate is not nil and error is nil
 	podInfo, err := cns.NewPodInfoFromIPConfigRequest(req)
 	if err != nil {
-		return cns.PodIpInfo{}, errors.Wrapf(err, "failed to parse IPConfigRequest %v", req)
+		return []cns.PodIpInfo{}, errors.Wrapf(err, "failed to parse IPConfigRequest %v", req)
 	}
 
 	if podIPInfo, isExist, err := service.GetExistingIPConfig(podInfo); err != nil || isExist {
 		return podIPInfo, err
 	}
 
 	// return desired IPConfig
-	if req.DesiredIPAddress != "" {
-		return service.AssignDesiredIPConfig(podInfo, req.DesiredIPAddress)
+	if req.DesiredIPAddresses != nil && len(req.DesiredIPAddresses) != 0 {
+		return service.AssignDesiredIPConfigs(podInfo, req.DesiredIPAddresses)
 	}
 
 	// return any free IPConfig
-	return service.AssignAnyAvailableIPConfig(podInfo)
+	return service.AssignAvailableIPConfigs(podInfo)
+	// TODO: create a check for returning a slice of PodIpInfo that is not full (i.e. slice of size two with only 1 IP)
```

Comment:
Can you pls add more context as to why we shouldn't return partial success? In that case, we rely on client to release the reserved IPs otherwise we will end up in a situation where we may reserve IPs w/o containers using those